source/extensions/common/wasm/context.h

```diff
@@ -450,6 +463,9 @@ class Context : public proxy_wasm::ContextBase,
   // Filter state prototype declaration.
   absl::flat_hash_map<std::string, Filters::Common::Expr::CelStatePrototypeConstPtr>
       state_prototypes_;
+
+  const bool no_automatic_route_refresh_{
```

Comment:
Changing the default behavior will break existing Wasm users (which is "beta" in Istio). I think we need to change the xDS API instead of the runtime key. Runtime key assumes the default behavior will eventually change but we can't be certain in this case.

Reply:
I do don't want break exist plugin, but I cannot find better chance to bring this back to the expected way. At least for now, the wasm still not production ready.

I also doubt if it is good choice to add new API. I think the API should be part of the wasm plugin configuration, not for the host configuration.

There are also some possible compromised way:
1. Add a runtime key (not runtime flag) to control the global default value.
2. Try to bump the ABI version and only change the behaviour in the new ABI version.

Reply:
Per offline discussion, a long-living runtime key is probably better to avoid a churn in the spec. Not refreshing routes with a custom host call is a safer option, until we get better guidance from Proxy-Wasm

Reply:
> I do don't want break exist plugins, but I cannot find better chance to bring this back to the expected way.

Don't change the default, and instead add xDS configuration option to disable automatic refresh.

I don't understand why you insisting on changing the default behavior and breaking existing plugins in the process.

> At least for now, the wasm still not production ready.

That's a bit of a stretch. People have been using Proxy-Wasm in Envoy in production for the last 4+ years.

> I also doubt if it is good choice to add new API. I think the API should be part of the wasm plugin configuration, not for the host configuration.

It absolutely should be a new xDS API:
1. This must be configurable on a per plugin basis (e.g. you might want to allow some plugins to clear route cache, but disable it for others).
2. As mentioned number of times already, since this is a security feature, hosts must be able to control that.

> Per offline discussion, a long-living runtime key is probably better to avoid a churn in the spec. Not refreshing routes with a custom host call is a safer option, until we get better guidance from Proxy-Wasm

To be honest, this is mostly Envoy internal change. However, the upcoming ABI v0.3 will add features (https://github.com/proxy-wasm/spec/issues/55) and ability to list custom functions (https://github.com/proxy-wasm/spec/issues/71), both of which could be used to let plugin know about changed behavior and addition of the custom hostcall.

Reply:
> That's a bit of a stretch. People have been using Proxy-Wasm in Envoy in production for the last 4+ years.

I am not very sure. May Proxy-Wasm self was production ready for a while. But in the Envoy, basically this is common understanding in our inner discussion (and it is still alpha). I think our recent target is to make it production ready in the Envoy. We just fixed a very very very basic bug #36411. So I persaonlly think may it's never be widely/heavily used until now.

> Don't change the default, and instead add xDS configuration option to disable automatic refresh.
> I don't understand why you insisting on changing the default behavior and breaking existing plugins in the process.

Because **implicitly** behavior will bring lots complexity in production where multiple different plugins are used. It require the developers know all the details or it may result in unexpected result.

So I think the correct way should be: 1) no implicitly refresh, 2) provide explict host call to do it, 3) provide additional control by the xDS if needed (mostly for security, although I think now it not a big problem, but this may be necessary in future once the wasm is widely used).

And considering current situation of wasm support of Envoy, I think this is the last timepoint to change the behavior and bring to the expected way.

Reply:
Of course, I also don't want to break exist plugin. So, I think we could provide different behavior for different versions' plugins.

 For example, only disable the automotical refreh for plugins with ABI version >= 0.3.0. Then all exist plugins won't be effected.

Reply:
> I am not very sure. May Proxy-Wasm self was production ready for a while. But in the Envoy, basically this is common understanding in our inner discussion (and it is still alpha). I think our recent target is to make it production ready in the Envoy. [...] So I persaonlly think may it's never be widely/heavily used until now.

AFAIK, it's "alpha" only because of the missing docs.

There are many people using it in production, but if you know better, then that's fine.

> We just fixed a very very very basic bug #36411. 

I've just looked at that bug and I'm a bit puzzled by it.

There were already tests for buffered and unbuffered responses, so it seems that Envoy behaves differently depending on HTTP version / codec? I'm not sure if that's desirable behavior. Also, did you check when this issue first appeared? Is it possible that it was a regression due to change from nghttp2 to oghttp2? Maybe that would explain why existing users didn't run into it yet?

> Of course, I also don't want to break exist plugin. So, I think we could provide different behavior for different versions' plugins.
> 
> For example, only disable the automotical refreh for plugins with ABI version >= 0.3.0. Then all exist plugins won't be effected.

Yes, that would be more reasonable. However, I still think this should be configurable in xDS.

Reply:
> There were already tests for buffered and unbuffered responses, so it seems that Envoy behaves differently depending on HTTP version / codec? I'm not sure if that's desirable behavior. Also, did you check when this issue first appeared? Is it possible that it was a regression due to change from nghttp2 to oghttp2? Maybe that would explain why existing users didn't run into it yet?

Envoy never promise that the last chunk will be empty. Even for the HTTP/1.1, the last chunk may be modified by filters and contains data.

So, it actually is a bug. But may be the lib transferation changes the possibility of the bug occurs? Not sure.
And if you look at that fix more carefully, you may found it also fixed another simple bug: not set the response buffer correctly.

Reply:
> Yes, that would be more reasonable. However, I still think this should be configurable in xDS.

I think we should make sure the default behavior match most cases and additional xds should only be used to provide additional control. And the ABI version also provide a way let's us to change the default behavior and without break exist plugins.

Anyway, it's hard to let everyone happy. Different people has different view. 

But seems this one could be an acceptable approach for our both.